File: 01_GOF23/DP_examples/C/03_Behavioral/Visitor/concrete_element.c
#include "concrete_element.h"
#include "visitor.h"
#include <stdio.h>

// Keyboard 实现
static void Keyboard_accept(IComputerPart* self, IComputerPartVisitor* visitor) {
    visitor->visitKeyboard(visitor, (struct Keyboard*)self);
}

void Keyboard_init(Keyboard* self) {
    self->base.accept = Keyboard_accept;
}

void Keyboard_operation(Keyboard* self) {
    printf("Using Keyboard\n");
}

// Monitor 实现
static void Monitor_accept(IComputerPart* self, IComputerPartVisitor* visitor) {
    visitor->visitMonitor(visitor, (struct Monitor*)self);
}

void Monitor_init(Monitor* self) {
    self->base.accept = Monitor_accept;
}

void Monitor_operation(Monitor* self) {
    printf("Using Monitor\n");
}

// Mouse 实现
static void Mouse_accept(IComputerPart* self, IComputerPartVisitor* visitor) {
    visitor->visitMouse(visitor, (struct Mouse*)self);
}

void Mouse_init(Mouse* self) {
    self->base.accept = Mouse_accept;
}

void Mouse_operation(Mouse* self) {
    printf("Using Mouse\n");
}

// Computer 实现
static void Computer_accept(IComputerPart* self, IComputerPartVisitor* visitor) {
    Computer* computer = (Computer*)self;
    for (int i = 0; i < 3; i++) {
        computer->parts[i]->accept(computer->parts[i], visitor);
    }
    visitor->visitComputer(visitor, computer);
}

void Computer_init(Computer* self) {
    self->base.accept = Computer_accept;
    
    // 初始化部件
    static Keyboard keyboard;
    static Mouse mouse;
    static Monitor monitor;
    
    Keyboard_init(&keyboard);
    Mouse_init(&mouse);
    Monitor_init(&monitor);
    
    self->parts[0] = (IComputerPart*)&keyboard;
    self->parts[1] = (IComputerPart*)&mouse;
    self->parts[2] = (IComputerPart*)&monitor;
}

File: 01_GOF23/DP_examples/Rust/03_Behavioral/Visitor/ConcreteElement.rs
// 具体元素：Keyboard

use crate::ObjectStructure::Computer;


pub struct Keyboard;

// 具体元素：Monitor
pub struct Monitor;

// 具体元素：Mouse
pub struct Mouse;

// 元素接口
pub trait IComputerPart {
    fn accept(&self, visitor: &dyn IComputerPartVisitor);
}

// 访问者接口
pub trait IComputerPartVisitor {
    fn visit_computer(&self, computer: &Computer);
    fn visit_mouse(&self, mouse: &Mouse);
    fn visit_keyboard(&self, keyboard: &Keyboard);
    fn visit_monitor(&self, monitor: &Monitor);
}

// 具体元素：Keyboard 实现
impl IComputerPart for Keyboard {
    fn accept(&self, visitor: &dyn IComputerPartVisitor) {
        visitor.visit_keyboard(self);
    }
}

impl Keyboard {
    pub fn operation(&self) {
        println!("Using Keyboard");
    }
}

// 具体元素：Monitor 实现
impl IComputerPart for Monitor {
    fn accept(&self, visitor: &dyn IComputerPartVisitor) {
        visitor.visit_monitor(self);
    }
}

impl Monitor {
    pub fn operation(&self) {
        println!("Using Monitor");
    }
}

// 具体元素：Mouse 实现
impl IComputerPart for Mouse {
    fn accept(&self, visitor: &dyn IComputerPartVisitor) {
        visitor.visit_mouse(self);
    }
}

impl Mouse {
    pub fn operation(&self) {
        println!("Using Mouse");
    }
}
